### PR TITLE
[ML-5700] Update release script to build and push docs.

### PR DESCRIPTION
**Pull request by MrBago**
This PR adds doc building to the release script. The script will checking the release tag and build the docs in docker. It will then push the docs branch, eg "gh-pages-1.4.0" upstream & replace the "gh-pages" branch.

I opted for replacing the gh-pages branch instead of creating a PR because this type of PR will inherently have conflicts.

### CODE REVIEW

File: dev/build-docs.sh [hunk 1]
```diff
@@ -10,4 +10,5 @@ docker build -t databricks/sparkdl-docs docs/
 build/sbt assembly
 
 # build the API docs
-docker run --rm -v "$(pwd):/mnt/sparkdl" databricks/sparkdl-docs dev/build-docs-in-docker.sh
+docker run --rm -v "$(pwd):/mnt/sparkdl" databricks/sparkdl-docs \
+    bash -i -c dev/build-docs-in-docker.sh
```
Comment (mengxr) [inline dev/build-docs.sh]: Shall we make a separate PR for this fix?

Comment (MrBago) [reply dev/build-docs.sh]: I can if you think we should, it's tiny so I was just folding it in.

File: dev/release.py [hunk 2]
```diff
@@ -4,9 +4,12 @@
 from subprocess import check_call, check_output
 import sys
 
-WORKING_BRANCH = "PREP_RELEASE_%s"
 DATABRICKS_REMOTE = "git@github.com:databricks/spark-deep-learning.git"
-PUBLISH_MODES = {"local": "publishLocal", "m2": "publishM2", "spark-package": "spDist"}
+PUBLISH_MODES = {"local": "publishLocal", "m2": "publishM2", "spark-package-local": "spDist",
+                 "spark-package-publish": "sbPublish"}
```
[the remark this answers is not in the document]
Comment (MrBago) [reply dev/release.py]: Sorry I don't quite follow, I think we need `spPublish` publishing to spark-packages, I can drop spDist.

File: dev/release.py [hunk 3]
```diff
@@ -60,9 +65,17 @@ def main(release_version, next_version, publish_to, no_prompt):
         sys.exit(1)
 
     working_branch = WORKING_BRANCH % release_version
-    if working_branch in check_output(["git", "branch"]):
-        prominentPrint(
-            "Working branch %s already exists, please delete it and try again." % working_branch)
+    gh_pages_branch = GH_PAGES_BRANCH % release_version
+    local_branches = check_output(["git", "branch"])
```
[the remark this answers is not in the document]
Comment (MrBago) [reply dev/release.py]: I updated the script, it works with python 2 & 3 now.

File: dev/release.py [hunk 4]
```diff
@@ -84,8 +98,30 @@ def main(release_version, next_version, publish_to, no_prompt):
...
+    check_call(["git", "checkout", "-b", gh_pages_branch, release_tag])
+    check_call(["./dev/build-docs.sh"])
+
+    commit_message = "Build docs for release %s." % release_version
+    check_call(["git", "add", "-f", "docs/_site"])
+    check_call(["git", "commit", "-m", commit_message])
+    msg = "Would you like to push docs branch to remote, %s, and update gh-pages branch? (y/n)"
+    msg %= git_remote
+    if verify(msg, interactive):
+        check_call(["git", "push", git_remote, gh_pages_branch])
```
[the remark this answers is not in the document]
Comment (MrBago) [reply dev/release.py]: Ya this is for quick revert in the future, but actually I think it should be a tag not a branch. I've made that change.

[the remark this answers is not in the document]
Comment (MrBago) [reply dev/release.py]: The problem with using a tag is that it shows up in the "releases" section of the github ui. I'll work on this tomorrow.

File: dev/release.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mengxr) [reply dev/release.py]: It was "sbPublish" instead of "spPublish". My ask is mainly why it shows up in a PR that is titled "doc build"?

File: dev/release.py [hunk 5]
```diff
@@ -1,12 +1,18 @@
...
 import click
+from datetime import datetime
 from six.moves import input
 from subprocess import check_call, check_output
 import sys
 
-WORKING_BRANCH = "PREP_RELEASE_%s"
 DATABRICKS_REMOTE = "git@github.com:databricks/spark-deep-learning.git"
-PUBLISH_MODES = {"local": "publishLocal", "m2": "publishM2", "spark-package": "spDist"}
+PUBLISH_MODES = {"local": "publishLocal", "m2": "publishM2", "spark-package-publish": "spPublish"}
```
Comment (mengxr) [inline dev/release.py]: I still don't know why the changes are relevant to "build and push docs". If they are necessary to include, it is worth an inline GitHub comment here. If they are irrelevant, we should create a separate PR.

Keeping PRs minimal is to help both PR review and merge. For example, I spent ~10 mins yesterday trying to understand how it is relevant to doc build but got no clue.

File: dev/release.py [hunk 6]
```diff
@@ -47,22 +56,33 @@ def main(release_version, next_version, publish_to, no_prompt):
...
         print(uncommitted_changes)
         print("There seem to be uncommitted changes on your current branch. Please commit or "
               "stash them and try again.")
         sys.exit(1)
 
-    working_branch = WORKING_BRANCH % release_version
-    if working_branch in check_output(["git", "branch"]):
-        prominentPrint(
-            "Working branch %s already exists, please delete it and try again." % working_branch)
+    working_branch = WORKING_BRANCH % (release_version, time.hour, time.minute, time.second)
```
Comment (mengxr) [inline dev/release.py]: It is missing date. It might be easier convert the timestamp into a single string first and include it in the template here.

File: dev/release.py [hunk 7]
```diff
@@ -47,22 +56,33 @@ def main(release_version, next_version, publish_to, no_prompt):
...
+    gh_pages_branch = WORKING_DOCS_BRANCH % (release_version, time.hour, time.minute, time.second)
```
Comment (mengxr) [inline dev/release.py]: Ditto.

File: dev/release.py (hunk 5 above)
Comment (MrBago) [reply dev/release.py]: This isn't related to docs, I should have commented sooner.

I realized while working on docs that we'll need this publish method available to use this for an actual release. I can split this into another PR if you think we should.

File: dev/release.py (hunk 6 above)
Comment (MrBago) [reply dev/release.py]: We need to use a custom format for the datetime because the default has characters that cannot be used in a git branch name, I'll update the custom format to include the date.

**Comment by MrBago** [comment]
@mengxr I think this is ready for another review, I'm also working on a PR to update the _config.yml file as part of the release. I tried changing `docs/_plugins/copy_api_dirs.rb` to use `version.sbt` instead of the config file, but it seems that jekyll reads the config file too, so that only updated some of the version references but not all.

**Comment by mengxr** [comment]
The doc release script looks good. I'm merging this PR in. However, there might be an issue with doc build. I cannot find the public APIs in the generated Python API html page. Could you see if this happens on your laptop?

**Comment by mengxr** [comment]
This is the error I hit:

~~~
/mnt/sparkdl/python/docs/index.rst:15: WARNING: toctree contains reference to nonexisting document u' sparkdl.graph'
/mnt/sparkdl/python/docs/index.rst:15: WARNING: toctree contains reference to nonexisting document u' sparkdl.image'
/mnt/sparkdl/python/docs/index.rst:15: WARNING: toctree contains reference to nonexisting document u' sparkdl.transformers'
/mnt/sparkdl/python/docs/index.rst:15: WARNING: toctree contains reference to nonexisting document u' sparkdl.udf'
/mnt/sparkdl/python/docs/index.rst:15: WARNING: toctree contains reference to nonexisting document u' sparkdl.utils'
WARNING: autodoc: failed to import module u'sparkdl'; the following exception was raised:
No module named tensorframes
WARNING: autodoc: failed to import module u'graph.input' from module u'sparkdl'; the following exception was raised:
No module named tensorframes
WARNING: autodoc: failed to import module u'graph.utils' from module u'sparkdl'; the following exception was raised:
No module named tensorframes
WARNING: autodoc: failed to import module u'image.imageIO' from module u'sparkdl'; the following exception was raised:
No module named tensorframes
WARNING: autodoc: failed to import module u'transformers.utils' from module u'sparkdl'; the following exception was raised:
No module named tensorframes
WARNING: autodoc: failed to import module u'udf.keras_image_model' from module u'sparkdl'; the following exception was raised:
No module named tensorframes
WARNING: autodoc: failed to import module u'utils.keras_model' from module u'sparkdl'; the following exception was raised:
No module named tensorframes
~~~
